modules/core/src/main/scala/doobie/util/write.scala

```diff
@@ -98,29 +98,22 @@ object Write extends WritePlatform with WriteLowerPriorityImplicits {
     new Write(
       puts = List((P, NoNulls)),
       toList = a => List(a),
-      unsafeSet = (ps, n, a) => P.unsafeSetNonNullable(ps, n, a),
-      unsafeUpdate = (rs, n, a) => P.unsafeUpdateNonNullable(rs, n, a),
-      unsafeSetOption = (ps, n, oa) => P.unsafeSetNullable(ps, n, oa),
-      unsafeUpdateOption = (rs, n, oa) => P.unsafeUpdateNullable(rs, n, oa),
-    )
-
-  implicit def fromPutOption[A](implicit P: Put[A]): Write[Option[A]] =
-    new Write(
-      puts = List((P, Nullable)),
-      toList = a => List(a),
-      unsafeSet = (ps, n, a) => P.unsafeSetNullable(ps, n, a),
-      unsafeUpdate = (rs, n, a) => P.unsafeUpdateNullable(rs, n, a),
-      unsafeSetOption = (ps, n, oa) => P.unsafeSetNullable(ps, n, oa.flatten),
-      unsafeUpdateOption = (rs, n, oa) => P.unsafeUpdateNullable(rs, n, oa.flatten),
+      unsafeSet = P.unsafeSetNonNullable,
+      unsafeUpdate = P.unsafeUpdateNonNullable,
+      unsafeSetOption = P.unsafeSetNullable,
+      unsafeUpdateOption = P.unsafeUpdateNullable,
     )
 }
 
 sealed trait WriteLowerPriorityImplicits {
   implicit def opt[A](implicit A: Write[A]): Write[Option[A]] = new Write[Option[A]](
     puts = A.puts.map { case (p, _) => (p, Nullable) },
-    toList = oa => oa.map(A.toList).getOrElse(A.puts.map(_ => null)),
-    unsafeSet = (ps, i, a) => A.unsafeSetOption(ps, i, a),
-    unsafeUpdate = (ps, i, a) => A.unsafeUpdateOption(ps, i, a),
+    toList = {
+      case None => A.puts.map(_ => None)
+      case Some(a) => A.toList(a).map(Some(_))
```

Comment:
Hm. So if we have an `Option` field in a case class, and we wrap that case class in another `Option`, then we'll have `Some(Some(...))` here (possibly up to unlimited depth)? I'm not sure if this is correct.

Reply:
We have a pattern-match [here](https://github.com/tpolecat/doobie/blob/d7c42833cf5aa3d8c26716d33ac5f76c1a5c531b/modules/core/src/main/scala/doobie/util/write.scala#L69-L70) that provides some evidence as to what the contract of `toList` is (a raw type element for each `NoNulls` column, an `Option` for each `Nullable`). Based on what happens to `puts` above, I'd say we want something like this:
```suggestion
      case Some(a) =>
        A.puts.zip(A.toList(a)).map {
          case ((_, NoNulls), x) => Some(x)
          case ((_, Nullable), ox) => ox
        }
```

Reply:
> Hm. So if we have an Option field in a case class, and we wrap that case class in another Option, then we'll have Some(Some(...)) here (possibly up to unlimited depth)

Yes, this exactly was my understanding when I wrote this code. 


Thanks for the link and the suggestion. Based on the pattern match you linked I think I convinced myself that it's correct. so I'll accept it and see the tests agree :)

Of course it would be preferable to capture this relationship in a type, and/or to produce some edge case tests and documentation for this thing. Any ideas in that direction @nigredo-tori ?

Reply:
Regarding capturing the relationship in a type - there might a way to capture that with `HList`s, but it won't be easy, and it will certainly require an overhaul of `Write`.

Regarding tests - I think that adding a type like this
```scala
class Foo(
  int: Int,
  intOpt: Option[Int]
)
```
, writing `Foo`, `Option[Foo]`, `Option[Option[Foo]]` to the database and reading them back would be a good enough test for this particular instance. For a more general approach, we can `zip` `puts`  from an arbitrary `Write` instance with its `toList` output, and check that we have `Option`s for (and only for) `Nullable` columns, and that we don't have nested options.